src/udp_mux/mod.rs

```diff
@@ -0,0 +1,332 @@
+/// address.
+fn normalize_socket_addr(target: &SocketAddr, socket_addr: &SocketAddr) -> SocketAddr {
+    match (target, socket_addr) {
+        (SocketAddr::V4(target_ipv4), SocketAddr::V6(_)) => {
+            let ipv6_mapped = target_ipv4.ip().to_ipv6_mapped();
+
+            SocketAddr::new(std::net::IpAddr::V6(ipv6_mapped), target_ipv4.port())
+        }
+        // This will fail later if target is IPv6 and socket is IPv4, we ignore it here
+        (_, _) => *target,
+    }
+}
+
+#[async_trait]
+pub trait UDPMux {
+    /// Close the muxing.
+    async fn close(&self);
+
+    /// Get the underlying connection for a given ufrag.
+    async fn get_conn(self: Arc<Self>, ufrag: &str) -> Result<Arc<dyn Conn + Send + Sync>, Error>;
+
+    /// Remove the underlying connection for a given ufrag.
+    async fn remove_conn_by_frag(&self, ufrag: &str);
+}
+
+#[derive(Debug)]
+pub struct UDPMuxParams {
+    udp_socket: UdpSocket,
+}
+
+impl UDPMuxParams {
+    pub fn new(udp_socket: UdpSocket) -> Self {
+        Self { udp_socket }
+    }
+}
+
+#[derive(Debug)]
+pub struct UDPMuxDefault {
+    /// The params this instance is configured with.
+    /// Contains the underlying UDP socket in use
+    params: UDPMuxParams,
+
+    /// Maps from ufrag to the underlying connection.
+    conns: Mutex<HashMap<String, UDPMuxConn>>,
+
+    /// Maps from ip address to the underlying connection.
+    address_map: RwLock<HashMap<SocketAddr, UDPMuxConn>>,
+
+    /// Whether this connection has been closed
+    closed: AtomicBool,
+}
+
+impl UDPMuxDefault {
+    pub fn new(params: UDPMuxParams) -> Arc<Self> {
+        let mux = Arc::new(Self {
+            params,
+            conns: Mutex::default(),
+            address_map: RwLock::default(),
+            closed: AtomicBool::new(false),
+        });
+
+        let cloned_mux = Arc::clone(&mux);
+        cloned_mux.start_conn_worker();
+
+        mux
+    }
+
+    pub fn is_closed(&self) -> bool {
+        self.closed.load(Ordering::SeqCst)
+    }
+
+    async fn send_to(&self, buf: &[u8], target: &SocketAddr) -> Result<usize, Error> {
+        self.params
+            .udp_socket
+            .send_to(buf, target)
+            .await
+            .map_err(Into::into)
+    }
+
+    /// Create a muxed connection for a given ufrag.
+    async fn create_muxed_conn(self: &Arc<Self>, ufrag: &str) -> UDPMuxConn {
+        let local_addr = self.params.udp_socket.local_addr().ok();
+
+        let params = UDPMuxConnParams {
+            local_addr,
+            key: ufrag.into(),
+            udp_mux: Arc::clone(self),
+        };
+
+        UDPMuxConn::new(params)
+    }
+
+    fn register_conn_for_address(&self, conn: &UDPMuxConn, addr: SocketAddr) {
+        if self.is_closed() {
+            return;
+        }
+
+        let key = conn.key();
+        {
+            let mut addresses = self
+                .address_map
+                .write()
+                .expect("Failed to obtain write lock");
```

Comment:
no expect() in library implementation, otherwise, panic may happen.

let's ->Result<()> in this function.

Reply:
This relates to this question we discussed about using locks from `std::sync`. If we migrate to the `parking_lot`'s [implementation](https://docs.rs/parking_lot/0.11.2/parking_lot/type.RwLock.html) this will go away. I was thinking of delaying this until a later PR, but I can do it in this PR if you wish